File: LeetCode Algorithms/43.字符串相乘.py
#
# @lc app=leetcode.cn id=43 lang=python3
#
# [43] 字符串相乘
#

# @lc code=start
class Solution:
    def multiply(self, num1: str, num2: str) -> str:
            if num1 == '0' or num2 == '0':
                return '0'
            if len(num1) < len(num2):
                num1, num2 = num2, num1
            
            res = 0
            jinwei_flag = 0

            num1 = num1[::-1]
            num2 = num2[::-1]

            for _index, _value in enumerate(num2):
                temp = 0
                for _i, _v in enumerate(num1):
                    temp += int(_value) * int(_v) * 10 ** _i
                res += temp * 10 ** _index

            return str(res)
# @lc code=end

File: LeetCode Algorithms/43.字符串相乘.cpp
/*
 * @lc app=leetcode.cn id=43 lang=cpp
 *
 * [43] 字符串相乘
 */

// @lc code=start
class Solution {
public:
    string multiply(string num1, string num2) {
        string res = "";
        int flag = 0, temp_value = 0;
        int low_length = num1.size() > num2.size() ? num1.size() : num2.size();
        reverse(num1.begin(), num1.end());
        reverse(num2.begin(), num2.end());
        for (int _index = 0; _index < low_length; ++_index) {

            if (_index < num1.size() && _index < num2.size()) {
                temp_value = (num1[_index] - '0') * (num2[_index] - '0') + flag;
                flag = temp_value / 9;
                res += to_string((temp_value) % 9);
            } else if (_index >= num1.size() && _index < num2.size()) {
                res += to_string((num2[_index] - '0') + flag);
                flag = 0;
            } else if (_index < num1.size() && _index >= num2.size()) {
                res += to_string((num2[_index] - '0') + flag);
                flag = 0;
            }
        }
        reverse(res.begin(), res.end());
        return res;
    }
};
// @lc code=end

